### PR TITLE
chore: Add pre-commit

### LINKED ISSUE #61: Add precommit hook for mypy and black
Add a simple pre-commit hook for `mypy` and `black` to avoid pushing code with errors that are quick and easy to fix.

### PR DESCRIPTION
**Pull request by edgarrmondragon**
Add pre-commit configuration for

- black
- mypy
- isort
- pyugrade (to ensure "newer" Python syntax is used)
- misc. checks (yaml, toml, json, whitespace)

Configuration for black, isort and mypy lives in `pyproject.toml`.

Closes #61

### CODE REVIEW

**Comment by edgarrmondragon** [comment]
I wonder if we should add a https://github.com/marketplace/pre-commit-ci job

File: pyproject.toml [hunk 1]
```diff
@@ -22,6 +22,20 @@ black = {version = "^21.5b1", allow-prereleases = true}
 types-requests = "^2.25.6"
 requests-cache = "^0.8.1"
 
+[[tool.mypy.overrides]]
+module = [
+    "backoff",
+]
+ignore_missing_imports = true
```
Comment (ericboucher) [inline pyproject.toml:29]: just curious, why is this needed?

**Review by ericboucher — APPROVED** [review]

**Comment by ericboucher** [comment]
> I wonder if we should add a [github.com/marketplace/pre-commit-ci](https://github.com/marketplace/pre-commit-ci) job

@edgarrmondragon is the pre-commit making the changes locally on fails? So that the dev just has to commit the changes to fix? I think that would be enough rather than having a pre-commit-ci :)

File: pyproject.toml (hunk 1 above)
Comment (edgarrmondragon) [reply pyproject.toml:29]: @ericboucher when a package like `backoff` is missing library stubs (such as `types-requests` for `requests`) or a `py.typed` file, mypy cannot get typing information from it.

This is the same as putting `# type: ignore` in front of any `import backoff` lines to make mypy happy.

**Comment by edgarrmondragon** [comment]
@ericboucher 

 > @edgarrmondragon is the pre-commit making the changes locally on fails?

That's correct. If a user runs `pre-commit install` then all the checks described above will run on every commit locally.

pre-commit-ci as a pipeline fails when pre-commit would make changes, so it's a way to not have users install pre-commit locally.

File: pyproject.toml (hunk 1 above)
Comment (ericboucher) [reply pyproject.toml:29]: So written like this, it only gets applied to the "backoff" module?

**Comment by ericboucher** [comment]
Let's start with this approach. You can go ahead and merge if the following comment is true - https://github.com/MeltanoLabs/tap-github/pull/65#discussion_r782447728
